### PR TITLE
feature: Add support for fewer braces

### PR DESCRIPTION
**Pull request by tgodzik**
More information: https://dotty.epfl.ch/docs/reference/other-new-features/indentation.html#variant-indentation-marker--for-arguments

### CODE REVIEW

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 1]
```diff
@@ -2146,6 +2153,36 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
     }
   }
 
+  private def followingIsLambdaAfterColon(): Boolean = {
+    /** Skip matching pairs of `(...)` or `[...]` parentheses.
+     *  @pre  The current token is `(` or `[`
+     */
+    def skipParens[T1: TokenClassifier, T2: TokenClassifier]: Unit = {
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:2160]: slightly rewritten, let me know if it's more readable:
```
def skipParens[T1: TokenClassifier, T2: TokenClassifier]: Unit = {
  // starts on left delimiter
  @tailrec
  def iter = {
    next()
    if (token.is[EOF]) false
    else if (token.is[T2]) true
    else  (token.isNot[T1] || skipParens[T1, T2]) && iter
  }
  // stops on right delimiter (true), or end (false)
  iter
}

ahead {
  val okParams = token match {
    case _: Ident | _: Underscore => true
    case _: LeftParen => skipParens[LeftParen, RightParen]
    case _: LeftBracket => skipParens[LeftBracket, RightBracket]
    case _ => false
  }
  okParams && {
    next()
    token.is[RightArrow] || token.is[ContextArrow]
  } && {
    next()
    token.is[Indentation.Indent]
  }
}
```

**Review by kitbellew — COMMENTED** [review]
lgtm. a few comments or suggestions.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 2]
```diff
@@ -2146,6 +2153,36 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
     }
   }
 
+  private def followingIsLambdaAfterColon(): Boolean = {
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:2156]: That's taken care of by checking the indentation token

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 1 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:2160]: Thanks! I looks better, I already rewrote it a bit from the Dotty version, but it was still lacking :sweat_smile:

**Comment by tgodzik** [comment]
> lgtm. a few comments or suggestions.

I realized, we need to handle infix also and I want to copy over some tests from dotty. Will continue tomorrow!

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 1 above)
Comment (kitbellew) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:2160]: could also do this, perhaps, without skipParens recursion:
```
def skipParens[T1: TokenClassifier, T2: TokenClassifier]: Unit = {
  // starts on left delimiter
  @tailrec
  def iter(nest: Int) = {
    next()
    if (token.is[EOF]) false
    else if (token.is[T2]) nest == 0 || iter(nest - 1)
    else (token.is[T1]) iter(nest + 1)
    else iter(nest)
  }
  // stops on right delimiter (true), or end (false)
  iter(0)
}
```

File: tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/FeverBracesSuite.scala [hunk 3]
```diff
@@ -0,0 +1,236 @@
+package scala.meta.tests.parsers.dotty
+
+import scala.meta.tests.parsers._
+import scala.meta._
+import scala.meta.tests.tokenizers.TokenizerSuite
+
+class FeverBracesSuite extends BaseDottySuite {
+
+  override implicit val dialect = dialects.Scala3.withAllowFewerBraces(true)
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/FeverBracesSuite.scala:9]: > I'm wondering if we are going to enable it by default for scala3.

We might, but I was thinking a safer approach would be to do it gradually.

> Otherwise we will have to add new options to scalafmt settings.

We can already change a single dialect options thanks to the changes by @kitbellew 

> Also, as it might be enabled by import we would need the same mechanic

We can enable it by default for nightly versions first and change it to stable later on.

[the remark this answers is not in the document]
Comment (dos65) [reply tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/FeverBracesSuite.scala:9]: > We can already change a single dialect options thanks to the changes by @kitbellew

oh, cool. didn't know that!

Then we would need only import mechanic but we can do it separate PR

[the remark this answers is not in the document]
Comment (dos65) [reply tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/FeverBracesSuite.scala:9]: I mean the case when you parse the source using std Scala3 dialect and if it contains:
`import import language.experimental.fewerBraces` then parser treats  it as  it was called with `Scala3.withAllowFewerBraces(true)` dialect

File: scalameta/dialects/shared/src/main/scala/scala/meta/Dialect.scala [hunk 4]
```diff
@@ -150,7 +150,9 @@ final class Dialect private (
     // Scala 3 uses proper precedence rules for infix types, unlike Scala 2
     val useInfixTypePrecedence: Boolean,
     // Scala213Source3 and Scala3 allow infix operator being placed after nl
-    val allowInfixOperatorAfterNL: Boolean
+    val allowInfixOperatorAfterNL: Boolean,
+    // Scala 3 will allow dropping braces for block arguments such as `list.map: a =>`
+    val allowFewerBraces: Boolean
```
Comment (kitbellew) [inline scalameta/dialects/shared/src/main/scala/scala/meta/Dialect.scala:155]: one additional point: currently, "fewer braces" is the name of the experimental import to enable it in the code but at some point it will be unnecessary, and then this name might become confusing. should we make it something like allowLambdaOptionalBraces?

Comment (tgodzik) [reply scalameta/dialects/shared/src/main/scala/scala/meta/Dialect.scala:155]: I think it should be good enough to keep the information in the comment and it will be easier for people currently to enable the dialect option. In the future, it might only be a problem for contributors, but the comment should suffice.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 5]
```diff
@@ -2138,6 +2138,13 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
           }
         }
         simpleExprRest(arguments, canApply = true, startPos = startPos)
+      case _: Colon if dialect.allowFewerBraces && (isColonEol(token) || followingIsLambdaAfterColon()) =>
+        accept[Colon]
+        in.observeIndented()
+        val args = blockExpr(allowRepeated = false)
+        val arguments = addPos { Term.Apply(t, List(args)) }
+        simpleExprRest(arguments, canApply = true, startPos = startPos)
+
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Removed

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 6]
```diff
@@ -2138,6 +2138,13 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
           }
         }
         simpleExprRest(arguments, canApply = true, startPos = startPos)
+      case _: Colon if dialect.allowFewerBraces && (isColonEol(token) || followingIsLambdaAfterColon()) =>
+        accept[Colon]
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Changed

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 7]
```diff
@@ -2146,6 +2153,36 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
...
+          next()
+          token.is[Indentation.Indent]
+        }
+      if (token.is[Ident] || token.is[Underscore]) {
+        next()
+        isArrowIndent()
+      } else if (token.is[LeftParen]) {
+        skipParens[LeftParen, RightParen]
+        isArrowIndent()
+
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Removed

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 8]
```diff
@@ -2138,6 +2138,13 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
           }
         }
         simpleExprRest(arguments, canApply = true, startPos = startPos)
+      case _: Colon if dialect.allowFewerBraces && (isColonEol(token) || followingIsLambdaAfterColon()) =>
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Done! Thanks for the suggestion!

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 1 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Good idea, I also added a test which has multiple `(` and `[`. I had to slightly rewrite TreeSyntax since it would always change to `{` if type was defined for parameter, but it's actually not needed since we already wrap it in `()` for all cases beside implicit and using

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 9]
```diff
@@ -2138,6 +2149,12 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
           }
         }
         simpleExprRest(arguments, canApply = true, startPos = startPos)
+      case _: Colon if allowFewerBraces =>
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:2152]: last question: do we need to check for `canApply` here (and is there a test which confirms that)?

**Review by kitbellew — APPROVED** [review]
one last comment below

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 9 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Added, it shouldn't break, but it should always be in an apply.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 10]
```diff
@@ -1637,60 +1648,6 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
             val contextFunction = token.is[ContextArrow]
             next()
             t = addPos {
-              def convertToParam(tree: Term): Option[Term.Param] = tree match {
```
Comment (tgodzik) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:1640]: Moved outside

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 11]
```diff
@@ -1550,7 +1554,14 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
...
             case _ =>
           }
+        } else if (token.is[Colon] && allowFewerBraces) {
+          next()
+          in.observeIndented()
+          val args = blockExpr(allowRepeated = false)
+          val arguments = addPos { Term.Apply(t, List(args)) }
+          t = simpleExprRest(arguments, canApply = true, startPos = startPos)
         } else if (acceptOpt[Colon]) {
+
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:1564]: extra empty line

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 12]
```diff
@@ -1716,6 +1672,61 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
...
+        )
+      )
+    case Lit.Unit() =>
+      None
+    case other =>
+      syntaxError(s"not a legal formal parameter", at = other)
+  }
+  private def convertToParams(tree: Term): List[Term.Param] = tree match {
+    case Term.Tuple(ts) => ts.toList flatMap convertToParam
+    case _ => List(convertToParam(tree)).flatten
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:1727]: or
```
  case Term.Tuple(ts) => ts.flatMap(convertToParam) // if tuple args are already a list; or toList at the end
  case _ => convertToParam(tree).toList // no need to flatten
```

File: scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala [hunk 13]
```diff
@@ -1140,9 +1140,11 @@ object TreeSyntax {
     // Multiples and optionals
     implicit def syntaxArgs: Syntax[List[Term]] = Syntax {
       case (b: Term.Block) :: Nil => s(" ", b)
-      case (f @ Term.Function(params, _)) :: Nil if !params.exists(_.decltpe.isEmpty) =>
+      case (f @ Term.Function(params, _)) :: Nil
+          if params.exists(_.mods.exists(m => m.is[Mod.Implicit] || m.is[Mod.Using])) =>
         s(" { ", f, " }")
-      case args => s("(", r(args, ", "), ")")
+      case args =>
+        s("(", r(args, ", "), ")")
```
Comment (kitbellew) [inline scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:1147]: what's the change here?

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 14]
```diff
@@ -2138,6 +2149,39 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
           }
         }
         simpleExprRest(arguments, canApply = true, startPos = startPos)
+      case _: Colon if canApply && allowFewerBraces =>
+        val isEol = isEolAfterColon(token)
+        next()
+        // map:
+        val args = if (isEol) {
+          in.observeIndented()
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:2157]: can this be
```
  map: /*
    line
    line */
   indentedCode
```

File: scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala [hunk 15]
```diff
@@ -1140,7 +1140,8 @@ object TreeSyntax {
     // Multiples and optionals
     implicit def syntaxArgs: Syntax[List[Term]] = Syntax {
       case (b: Term.Block) :: Nil => s(" ", b)
-      case (f @ Term.Function(params, _)) :: Nil if !params.exists(_.decltpe.isEmpty) =>
```
Comment (kitbellew) [inline scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:1143]: forgot to ask: why this change?

Comment (tgodzik) [reply scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:1143]: It was adding {} for `apply( (a: String) => a )` which is not neccessary and was breaking one of the added tests, since {} would be a different tree

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 14 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:2156]: Works!

File: scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala (hunk 13 above)
Comment (tgodzik) [reply scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala]: Removed

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 12 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Changed!

**Review by kitbellew — APPROVED** [review]

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 16]
```diff
@@ -2138,6 +2148,40 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
...
+          /**
+           * We need to handle param and then open indented region, otherwise only the block will be
+           * handles and any `.` will be accepted into the block:
+           * ```
+           * .map: a =>
+           *   a+1
+           * .filter: x =>
+           *   x > 2
+           * ```
+           * Witout manual handling here, filter would be included for `(a+1).filter`
```
Comment (dos65) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:2169]:
```suggestion
           * Without manual handling here, filter would be included for `(a+1).filter`
```

**Review by dos65 — APPROVED** [review]
:fire:

**Comment by tgodzik** [comment]
Now we need to handle it in `scalafmt` :sweat_smile: I will take a look, there is one more fix I want to include in the release.

**Comment by kitbellew** [comment]
> Now we need to handle it in `scalafmt` 😅 I will take a look, there is one more fix I want to include in the release.

and i have a couple of additional fixes on top of #2815.
